src/main/kotlin/anttracker/PageOf.kt

```diff
@@ -0,0 +1,119 @@
+            val queryOutput = queryToDB()
+            if (queryOutput != null) {
+                queryOutput.forEach { record ->
+                    records.add(record)
+                }
+            } else {
+                throw Exception("PageOf: Attempted to load from null query output.")
+            }
+        }
+    }
+
+    // -------------------------------------------------------------------------------
+    // Returns the size/length of the records MutableList.
+    // ---
+    fun recordsSize(): Int = records.size
+
+    // -------------------------------------------------------------------------------
+    // Getter for elements in the records MutableList.
+    // ---
+    fun getContentAt(index: Int): IntEntity = records[index]
+
+    // -------------------------------------------------------------------------------
+    // Increments page number and loads associated DB records into records MutableList
+    // Throws an error if attempting to load beyond the last page.
+    // ---
+    fun loadNextPage() {
+        if (lastPage()) {
+            throw Exception("PageOf: Already reached last page.")
+        }
+        pagenum++
+        loadRecords()
+    }
+
+    // -------------------------------------------------------------------------------
+    // TRUE if current page >= than the last page.
+    // ---
+    fun lastPage(): Boolean = pagenum >= lastPageNum
+
+    // -------------------------------------------------------------------------------
+    // Returns the number of yet-to-be-displayed records after the current page.
+    // ---
+    fun countRemainingRecords(): Int = getQueryRecordCount() - (queryLimit * (pagenum + 1))
+
+    // -------------------------------------------------------------------------------
+    // Used in init/ctor block to calculate the last page number.
+    // As we are 0-indexing the page numbers, we subtract 1 at the end.
+    // ---
+    protected fun initLastPageNum() {
+        lastPageNum =
+            ceil(
+                (getQueryRecordCount().toDouble() / queryLimit.toDouble()),
+            ).toInt() - 1
+    }
+
+    // -------------------------------------------------------------------------------
+    // Returns the total number of records/rows of the query being paginated.
+    // Used for calculating the number of pages the query needs.
+    // ---
+    private fun getQueryRecordCount(): Int {
+        var count: Int = -1
+        transaction {
+            val queryOutput: SizedIterable<IntEntity>? = queryToDB()
+            if (queryOutput != null) {
+                count = queryOutput.count().toInt()
+            }
+        }
+        return count
+    }
+
+    // -------------------------------------------------------------------------------
+    // Defines the DAO query to DB used to pull records into memory.
+    // Abstract/Virtual as query needs to be defined per PageOf Type
+    // ---
+    protected abstract fun queryToDB(): SizedIterable<IntEntity>?
```

Comment:
In what case would the query return `null`? Also, if `T` is the generic type, why not use that instead of `IntEntity`? Also, why are you returning a `SizedIterable` instead of an `Iterable`?

Reply:
Nit: I find that `queryToDB` does not convey what the function is doing. Something like `fetchRecords` or `fetchRecordsPage` would be a tad more clear.

Reply:
- queryToDB: agreed on the name, I like fetchRecords so I will change it to that.
- SizedIterable: that's what the .find function was returning so I was using that. Iterable works too but don't see a need. Same goes for IntEntity vs Entity
- query returning null: That was me being a little lazy to declare an iterable without defining it. If you can figure out a better/cleaner way to have a function return an Iterable from a transaction block I'm all ears.

Reply:
IMHO, not seeing a need hints that we should use a more generic type.

I replaced `queryToDB` to return non-null and it works fine. Nothing else to change on initialization and no need to check for null.

Reply:
now that I've realized you can return the transaction block I'm all for it. fixed